---
title: "Decoding neuronal diversity by single cell Convert-seq"
author: "Joachim"
date: "May 30, 2018"
output: html_document
---

##########################################################
################         LIBRARIES        ################
##########################################################

```{r LOAD LIBRARIES, warning = FALSE, message = FALSE}

pacman::p_load(readxl, edgeR, tidyverse, reshape, RColorBrewer, Seurat, rio, data.table, monocle, pheatmap, dendextend)

```

##########################################################
################      CUSTOM FUNCTIONS    ################
##########################################################

```{r CUSTOM FUNCTIONS, warning = FALSE, message = FALSE}
#--------------------------Read data------------------------------
readat <- function(id) {
  pid = paste0("C:/Users/joachim/Desktop/scRNAseq.iN.R/", id)
  dat = import(pid)
  dat = dat %>% remove_rownames %>% column_to_rownames(var = names(dat)[1])
}

readex <- function(id, sheet) {
  pid = paste0("C:/Users/joachim/Desktop/scRNAseq.iN.R/", id)
  dat = read_excel(pid, sheet = sheet)
}

#----------------------------Theme1--------------------------------
theme_jo <- function(base_size = 12){
  theme_classic(base_size = base_size) +
    theme(
      legend.key = element_blank(), 
      strip.background = element_blank(),
      strip.text = element_text(size = 20, color = "grey10"),
      plot.title = element_blank(), 
      axis.title = element_text(size = 30, color = "grey10"), 
      axis.text = element_text(size = 25, color = "grey10"), 
      legend.title = element_blank(), 
      legend.text = element_blank(),
      axis.line = element_line(size = 1.1, color = "grey10"),
      axis.ticks = element_line(size = 1.1, color = "grey10"),
      axis.ticks.length = unit(0.3, "cm"),
      legend.position = "none")
}

#----------------------------Theme2--------------------------------
theme_blank = function(base_size = 12){
  theme_jo(base_size = base_size) +
    theme(
      axis.text = element_blank(),
      axis.title = element_blank(),
      panel.background = element_blank(),
      strip.text = element_blank()
    )
}

#----------------------------Multiple replacements--------------------------------

gsub2 <- function(pattern, replacement, x, ...) {
  if (length(pattern)!=length(replacement)) {
    stop("pattern and replacement do not have the same length.")
  }
  result <- x
  for (i in 1:length(pattern)) {
    result <- gsub(pattern[i], replacement[i], result, ...)
  }
  result
}

#----------------------------Subchunkify--------------------------------

subchunkify <- function(g, fig_height = 7, fig_width = 5) {
  g_deparsed <- paste0(deparse(
    function() {g}
  ), collapse = '')
  
  sub_chunk <- paste0("
  `","``{r sub_chunk_", floor(runif(1) * 10000), ", fig.height=", fig_height, ", fig.width=", fig_width, ", echo=FALSE}",
  "\n(", 
    g_deparsed
    , ")()",
  "\n`","``
  ")
  
  cat(knitr::knit(text = knitr::knit_expand(text = sub_chunk), quiet = TRUE))
}
```

##########################################################
#################        FIGURE S1        ################
##########################################################

```{r FIGURE S1, warning = FALSE, message = FALSE, results = 'asis'}
#----------------------------FigureS1A--------------------------------
#read data
dat.ips = readex("ips.to.neuron.xlsx", 1) %>% column_to_rownames(var = "Gene") %>% as.data.frame
dat.nervous = readex("ips.to.neuron.xlsx", 2)
dat.cand = readex("ips.to.neuron.xlsx", 3) %>% mutate(p.Gene = paste0("p1@", .$Gene))
dat.tf = readex("ips.to.neuron.xlsx", 4)

#define groups and subset data
group = dat.ips %>% colnames %>% stringr::str_split("_") %>% sapply(., function(X) X[3]) %>% .[1:12]
f.data = dat.ips[rowSums(dat.ips) > 5, 1:12]

#combine data and group; Calculate dispersions; Fit dispersion
my.data = f.data %>% DGEList(counts = ., group = group) %>% estimateCommonDisp %>% estimateTagwiseDisp
design = model.matrix(~group)
colnames(design) = unique(group)
fit = glmFit(my.data, design)

#ANOVA-like analysis
lrt = glmLRT(fit, coef = 2:4)
lrt$table$FDR = p.adjust(lrt$table$PValue, method = "BH")
de.table = lrt$table

#set a cutoff, add day0
sig.data = de.table %>%
  rownames_to_column %>%
  filter(FDR <= 0.05) %>% 
  select(1:4) %>% 
  setNames(gsub("logFC.", "", names(.))) %>% 
  mutate(day00 = 0) %>% 
  column_to_rownames(var = "rowname") %>%
  select("day00", "day06", "day12", "day18") %>%
  t %>% 
  as.data.frame

#select TFs
tfs = dat.tf %>% 
  mutate(p.Gene = paste0("p1@", .$Gene))
sig.data.tf = sig.data[, which(colnames(sig.data) %in% tfs$p.Gene)] %>%
  t %>% 
  melt %>% 
  full_join(dat.cand, by = c("X1" = "p.Gene")) %>% 
  select(c("X1", "X2", "value", "Color"))
sig.data.tf = sig.data.tf[-(which(is.na(sig.data.tf$value) == "TRUE")), ]
sig.data.tf[is.na(sig.data.tf)] = "Z"

#create color palette
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
my.pal = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
my.pal = sample(my.pal)
my.pal.ips = c("Z" = "grey90", "A" = my.pal[1], "B" = my.pal[2],"C" = my.pal[3],"D" = my.pal[4],"E" = my.pal[5],
               "F" = my.pal[6],"G" = my.pal[7],"H" = my.pal[8],"I" = my.pal[9],"J" = my.pal[10],"K" = my.pal[11],
               "L" = my.pal[12],"M" = my.pal[13],"N" = my.pal[14],"O" = my.pal[15],"P" = my.pal[16],"Q" = my.pal[17],
               "R" = my.pal[18],"S" = my.pal[19],"T" = my.pal[20], "U" = "#525252", "V" = "#525252")

#ploy data
p = ggplot(sig.data.tf, aes(X2, value, group = X1, color = as.factor(Color), fill = as.factor(Color))) +
  geom_line(alpha = 0.5, lwd = 1, lineend = "round") +
  geom_point(data = sig.data.tf[-(which(sig.data.tf$Color == "Z")), ], size = 5, pch = 19) +
  geom_point(data = sig.data.tf[-(which(sig.data.tf$Color == "Z")), ], size = 3, pch = 19, color = "white") +
  geom_line(data = sig.data.tf[-(which(sig.data.tf$Color == "Z")), ], lwd = 1.5, lineend = "round") +
  scale_color_manual(values = my.pal.ips) +
  scale_fill_manual(values = my.pal.ips) +
  scale_x_discrete(expand = c(0.1, 0.1), labels = c("Day0", "Day6", "Day12", "Day18")) +
  labs(x = "Time-point", y = "Log2(FC vs. Day0)") +
  geom_hline(aes(yintercept = 0), color = "grey20", lty = 2, size = 1) +
  theme_jo() +
  theme(legend.position = "none")
subchunkify(p, 6, 8)
```

##########################################################
#################         FIGURE 2       #################
##########################################################

```{r FIGURE 2, warning = FALSE, message = FALSE, results = 'asis'}
#----------------------------Figure2A--------------------------------
#read 10x data, create Seurat
dat.ref.10x.count = Read10X("C:/Users/joachim/Desktop/scRNAseq.iN.R/TFi.Ci.10x.g19")

#seurat
iNS.10x = CreateSeuratObject(raw.data = dat.ref.10x.count, 
                             min.cells = 3, 
                             min.genes = 200, 
                             normalization.method = "LogNormalize", 
                             display.progress = F)

#regress on mitochondrial genes and number of UMI
mito.genes = grep(pattern = "MT-", x = rownames(x = iNS.10x@data), value = TRUE)
percent.mito = Matrix::colSums(iNS.10x@raw.data[mito.genes, ]) / Matrix::colSums(iNS.10x@raw.data)

#cluster cells
iNS.10x = AddMetaData(object = iNS.10x, metadata = percent.mito, col.name = "percent.mito") %>%
  FilterCells(subset.names = c("nGene", "percent.mito"), low.thresholds = c(1000, -Inf), high.thresholds = c(4500, 0.2)) %>%
  NormalizeData(display.progress = T) %>%
  FindVariableGenes(x.low.cutoff = 0.01, x.high.cutoff = 8, y.cutoff = 1, do.plot = F, display.progress = F) %>%
  ScaleData(vars.to.regress = c("nUMI", "percent.mito"), display.progress = F) %>%
  RunPCA(pcs.compute = 60, do.print = F, pcs.print = 0, genes.print = 0) %>%
  ProjectPCA(do.print = FALSE, pcs.print = 0, pcs.store = 60, genes.print = 0) %>%
  FindClusters(dims.use = 1:20, resolution = 0.7, print.output = F) %>%
  RunTSNE(dims.use = 1:20, do.fast = T)

#create data frame for ggplot
iNS.10x.df = as.data.frame(iNS.10x@dr$tsne@cell.embeddings) %>%
  rownames_to_column(var = "Cell") %>%
  mutate(Sample = factor(unlist(lapply(strsplit(Cell, "[.]"), function(X) X[[1]])), levels = c("TFi", "Ci")),
         Cluster = factor(iNS.10x@meta.data$res.0.7, levels = c(0, 2, 1, 6, 3, 5, 4, 7, 8)))

#create color palettes
my.pal.10x.tsne.fill = c("0" = "#063c8d", "1" = "#810081", "2" = "#0a8873", "3" = "#c51385", "4" = "#df0909", 
                         "5" = "#f55701", "6" = "#00919e", "7" = "#f4ea0b", "8" = "#00b900")    
my.pal.10x.tsne.col = c("0" = "#031b40", "1" = "#360036", "2" = "#043b32", "3" = "#780c50", "4" = "#910606",
                        "5" = "#a83c02", "6" = "#004b52", "7" = "#a8a307", "8" = "#006e00") 

#plot tSNE
p = ggplot(iNS.10x.df, aes(tSNE_1, tSNE_2, fill = Cluster, color = Cluster, 
                           shape = as.factor(Sample), size = as.factor(Sample))) +
  geom_point(alpha = 0.8) +
  scale_fill_manual(values = my.pal.10x.tsne.fill) +
  scale_color_manual(values = my.pal.10x.tsne.col) +
  scale_shape_manual(values = c(21, 24)) +
  scale_size_manual(values = c(5, 4)) +
  theme_jo()
subchunkify(p, 8, 8)

#----------------------------Figure2D--------------------------------
#create ref data
dat.ref.10x = as.matrix(iNS.10x@data) %>%
  cpm 
dat.ref.10x = log2(dat.ref.10x + 1)

#define marker gene
marker.genes.10x = c("SPOCK1","FRZB", "MKI67", "CENPA", "DRD4","SEMA3G","CHRNA5", "CP", "SLC6A17", "INSM1", "ANK3","NKAIN4") 
dat.ref.10x.m = as.data.frame(t(dat.ref.10x[marker.genes.10x, ])) %>%
  rownames_to_column(var = "Cell")

#create data frame
iNS.10x.df.m = iNS.10x.df %>%
  mutate(Cluster = factor(Cluster, levels = c(0, 2, 1, 3, 6, 8, 5, 7, 4))) %>%
  left_join(., dat.ref.10x.m, by = "Cell") %>%
  melt(id = c("Cell", "tSNE_1", "tSNE_2", "Sample", "Cluster")) %>%
  mutate(CPM = value,
         Gene = variable) %>%
  select(c("Cluster", "Gene", "CPM"))

#create color palettes
my.pal.10x.tsne.fill = c("0" = "#063c8d", "1" = "#810081", "2" = "#0a8873", "3" = "#c51385", "4" = "#df0909", 
                         "5" = "#f55701", "6" = "#00919e", "7" = "#f4ea0b", "8" = "#00b900")    
my.pal.10x.tsne.col = c("0" = "#031b40", "1" = "#360036", "2" = "#043b32", "3" = "#780c50", "4" = "#910606",
                        "5" = "#a83c02", "6" = "#004b52", "7" = "#a8a307", "8" = "#006e00") 

ggplot(iNS.10x.df.m, aes(Cluster, CPM, fill = Cluster, color = Cluster)) +
  geom_violin(scale = "width", lwd = 0.6, alpha = 1, trim = T) +
  ylab("log2(CPM)") +
  scale_color_manual(values = my.pal.10x.tsne.col) +
  scale_fill_manual(values = my.pal.10x.tsne.fill) +
  coord_flip()+
  scale_x_discrete(limits = rev(levels(iNS.10x.df.m$Cluster))) +
  theme_jo() +
  facet_wrap(~Gene, nrow = 1)

```

##########################################################
#################        FIGURE 4        #################
##########################################################

```{r FIGURE 4, warning = FALSE, message = FALSE, results = 'asis'}
#----------------------------Figure4B--------------------------------
#read data
dat.ref.tc.count = readat("time.course.count.mat.qc.csv")

#tSNE early
dat.ref.tc.count.Vc.iN1 = dat.ref.tc.count[, grep(c("FIB|Ci1|TFi1|TFi2"), colnames(dat.ref.tc.count))]
tc.data.Vc.iN1 = Matrix(as.matrix(dat.ref.tc.count.Vc.iN1), sparse = T)

#seurat
tcs.Vc.iN1 = CreateSeuratObject(raw.data = tc.data.Vc.iN1, min.cells = 3, normalization.method = "LogNormalize", display.progress = F) %>%
  FindVariableGenes(x.low.cutoff = 0.01, x.high.cutoff = 8, y.cutoff = 1, do.plot = F, display.progress = F) %>%
  ScaleData(display.progress = F) %>%
  RunPCA(do.print = F, pcs.print = 0, genes.print = 0) %>%
  ProjectPCA(do.print = F, pcs.print = 0, genes.print = 0) %>%
  FindClusters(dims.use = 1:7, resolution = 1, print.output = F) %>%
  RunTSNE(dims.use = 1:7, do.fast = T)

#create data frame
tsn.tc.Vc.iN1 = as.data.frame(tcs.Vc.iN1@dr$tsne@cell.embeddings) %>%
  rownames_to_column(var="Cell") %>%
  mutate(Sample = unlist(lapply(strsplit(Cell, "[.]"), function(X) X[[1]])),
         Cluster = factor(tcs.Vc.iN1@meta.data$res.1, levels = c(3, 4, 0, 1, 2)),
         Shape = factor(gsub("TFi2", "TFi1", Sample), levels = c("FIB", "Ci1", "TFi1")))

#create palettes
my.pal.tc.tsne.s = colorRampPalette(c("#004829", "white"))(6)
my.pal.tc.tsne.c1 = c("#033782", "#8b008b", "#d40b91", "#ec1717", "#fa7600", "#f4eb18")
my.pal.tc.tsne.c2 = c("#011736", "#400040", "#87075c", "#a11010", "#ad5100", "#8c820e")
my.pal.tc.tsne.s1 = c("grey20", "grey40", "grey60")
my.pal.tc.tsne.s2 = c("grey50", "grey70", "grey90")

#plot
p = ggplot(tsn.tc.Vc.iN1, aes(tSNE_1, tSNE_2, fill = Cluster, color = Cluster)) +
  geom_point(size = 6, shape = 21) +
  scale_fill_manual(values = my.pal.tc.tsne.c1) +
  scale_color_manual(values = my.pal.tc.tsne.c2) +
  theme_jo() +
  theme(panel.border = element_rect(fill = NA, size = 1.5, color = "grey10"),
        axis.line = element_blank(),
        axis.ticks = element_line(size = 1, color = "grey10"))
subchunkify(p, 6, 6)

#tSNE late
dat.ref.tc.count.Vc.iN2 = dat.ref.tc.count[, grep(c("FIB|Ci2|TFi3|TFi4"), colnames(dat.ref.tc.count))]
tc.data.Vc.iN2 = Matrix(as.matrix(dat.ref.tc.count.Vc.iN2), sparse = T)

#seurat
tcs.Vc.iN2 = CreateSeuratObject(raw.data = tc.data.Vc.iN2, min.cells = 3, normalization.method = "LogNormalize", display.progress = F) %>%
  FindVariableGenes(x.low.cutoff = 0.3, x.high.cutoff = 8, y.cutoff = 1, do.plot = F, display.progress = F) %>%
  ScaleData(display.progress = F) %>%
  RunPCA(do.print = F, pcs.print = 0, genes.print = 0) %>%
  ProjectPCA(do.print = F, pcs.print = 0, genes.print = 0) %>%
  FindClusters(dims.use = 1:5, resolution = 1, print.output = F) %>%
  RunTSNE(dims.use = 1:5, do.fast = T)

#create data frame
tsn.tc.Vc.iN2 = as.data.frame(tcs.Vc.iN2@dr$tsne@cell.embeddings) %>%
  rownames_to_column(var="Cell") %>%
  mutate(Sample = unlist(lapply(strsplit(Cell, "[.]"), function(X) X[[1]])),
         Cluster = factor(tcs.Vc.iN2@meta.data$res.1, levels = c(2, 3, 0, 1)),
         Shape = factor(gsub("TFi4", "TFi3", Sample), levels = c("FIB", "Ci2", "TFi3"))) 

my.pal.tc.tsne.c1 = c("#033782", "#8b008b", "#ec1717", "#fa7600", "#f4eb18")
my.pal.tc.tsne.c2 = c("#011736", "#400040", "#a11010", "#ad5100", "#8c820e")

#plot clusters
p = ggplot(tsn.tc.Vc.iN2, aes(tSNE_1, tSNE_2, fill = Cluster, color = Cluster)) +
  geom_point(size = 6, shape = 21) +
  scale_fill_manual(values = my.pal.tc.tsne.c1) +
  scale_color_manual(values = my.pal.tc.tsne.c2) +
  theme_jo() +
  theme(panel.border = element_rect(fill = NA, size = 1.5, color = "grey10"),
        axis.line = element_blank(),
        axis.ticks = element_line(size = 1, color = "grey10"))
subchunkify(p, 6, 6)
```

##########################################################
#################        FIGURE S4       #################
##########################################################

```{r FIGURE S4, warning = FALSE, message = FALSE, results = 'asis'}
#----------------------------FigureS4E-F--------------------------------
#create color palettes
my.pal.tc.batch1 = c("FIB" = "grey80", "Ci1" = "grey80", "Ci2" = "grey80", "TFi1" = "#eb1616", 
                     "TFi2" = "#f97600", "TFi3" = "#f0bc00", "TFi4" = "#5ac000")
my.pal.tc.batch2 = c("FIB" = "grey50", "Ci1" = "grey50", "Ci2" = "grey50", "TFi1" = "#9e0e0e", 
                     "TFi2" = "#ad5100", "TFi3" = "#a38000", "TFi4" = "#367300")
#plot
p = ggplot(tsn.tc.Vc.iN1, aes(tSNE_1, tSNE_2, fill = as.factor(Sample), color = as.factor(Sample))) +
  geom_point(size = 6, shape = 21) +
  scale_fill_manual(values = my.pal.tc.batch1) +
  scale_color_manual(values = my.pal.tc.batch2) +
  theme_jo()+
  theme(panel.border = element_rect(fill = NA, size = 1.5, color = "grey10"),
        axis.ticks = element_line(size = 1),
        axis.title = element_blank(),
        axis.line = element_blank())
subchunkify(p, 6, 6)

#plot
p = ggplot(tsn.tc.Vc.iN2, aes(tSNE_1, tSNE_2, fill = as.factor(Sample), color = as.factor(Sample))) +
  geom_point(size = 6, shape = 21) +
  scale_fill_manual(values = my.pal.tc.batch1) +
  scale_color_manual(values = my.pal.tc.batch2) +
  theme_blank() +
  theme(panel.border = element_rect(fill = NA, size = 1.5, color = "grey10"),
        axis.ticks = element_line(size = 1),
        axis.title = element_blank(),
        axis.line = element_blank())
subchunkify(p, 6, 6)

#----------------------------FigureS4G--------------------------------
#read TPM data
dat.tc.ref = readat("time.course.tpm.mat.qc.log.csv")

#dendrogram early
dat.tpm.iN1 = dat.tc.ref[which(row.names(dat.tc.ref) %in% tcs.Vc.iN1@var.genes), which(colnames(dat.tc.ref) %in% tsn.tc.Vc.iN1$Cell)]

#create dendrogram
dat.dend = t(dat.tpm.iN1) %>%
  scale %>%
  dist(method = "euclidean") %>%
  hclust(method = "ward.D2") %>%
  as.dendrogram

#modify dendrogram
dend.label = dat.dend %>%
  labels %>%
  as.data.frame
dend.annot = data.frame(Cluster = as.character(tsn.tc.Vc.iN1$Cluster)) %>%
  mutate(rownames = tsn.tc.Vc.iN1$Cell,
         Cluster = chartr("34012", "ABCDE", Cluster),
         Group = tsn.tc.Vc.iN1$Sample)
dend.col = left_join(dend.annot, dend.label, by = c("rownames" = ".")) %>%
  mutate(Color = gsub2(c("A", "B", "C", "D", "E"), c("#033782", "#8b008b", "#d40b91", 
                                                     "#ec1717", "#fa7600"), Cluster)) %>%
  column_to_rownames(var = "rownames")
dend.label = dend.label %>% column_to_rownames(var = ".")
dend.col = dend.col[row.names(dend.label), ] 
dend.ord = dend.col$Color

#plot dendrogram
dat.dend %>% 
  set("leaves_pch", 19) %>%
  set("leaves_cex", 1) %>%
  hang.dendrogram %>%
  set("leaves_col", dend.ord) %>%
  rotate(as.character(dend.annot[order(dend.annot$Cluster),]$rownames)) %>%
  set("labels", rep(NA, times=length(labels(.)))) %>%
  set("branches_lwd", 1.3) %>%
  set("branches_k_col", c("#033782", "#8b008b", "#d40b91", "#ec1717", "#fa7600"), k=5) %>%
  hang.dendrogram(hang_height = 30) %>%
  plot

#dendrogram late
dat.tpm.iN2 = dat.tc.ref[which(row.names(dat.tc.ref) %in% tcs.Vc.iN2@var.genes), which(colnames(dat.tc.ref) %in% tsn.tc.Vc.iN2$Cell)]

#create dendrogram
dat.dend = t(dat.tpm.iN2) %>%
  scale %>%
  dist(method = "euclidean") %>%
  hclust(method = "ward.D2") %>%
  as.dendrogram

#modify dendrogram
dend.label = dat.dend %>%
  labels %>%
  as.data.frame
dend.annot = data.frame(Cluster = as.character(tsn.tc.Vc.iN2$Cluster)) %>%
  mutate(rownames = tsn.tc.Vc.iN2$Cell,
         Cluster = chartr("2301", "ABCD", Cluster),
         Group = tsn.tc.Vc.iN2$Sample)
dend.col = left_join(dend.annot, dend.label, by = c("rownames" = ".")) %>%
  mutate(Color = gsub2(c("A", "B", "C", "D"), c("#033782", "#8b008b", 
                                                "#ec1717", "#fa7600"), Cluster)) %>%
  column_to_rownames(var = "rownames")
dend.label = dend.label %>% column_to_rownames(var = ".")
dend.col = dend.col[row.names(dend.label), ] 
dend.ord = dend.col$Color

#plot dendrogram
dat.dend %>% 
  set("leaves_pch", 19) %>%
  set("leaves_cex", 1) %>%
  hang.dendrogram %>%
  set("leaves_col", dend.ord) %>%
  rotate(as.character(dend.annot[order(dend.annot$Cluster),]$rownames)) %>%
  set("labels", rep(NA, times=length(labels(.)))) %>%
  set("branches_lwd", 1.3) %>%
  set("branches_k_col", c("#033782", "#8b008b", "#ec1717", "#fa7600"), k=4) %>%
  hang.dendrogram(hang_height = 15) %>%
  plot

#----------------------------FigureS4H--------------------------------
#define exogenous genes
marker.genes.exo = c("ASCL1_EXO", "DLX1_EXO", "DLX2_EXO", "FEV_EXO", "FOXA2_EXO", "FOXP2_EXO", "ISL1_EXO", 
                     "LHX2_EXO", "NEUROD1_EXO", "NEUROG2_EXO", "NR2F1_EXO", "NR2F2_EXO", "NR4A2_EXO", 
                     "OLIG2_EXO", "OTX2_EXO", "PAX6_EXO", "PITX3_EXO", "POU3F2_EXO", "TLX3_EXO", "ZIC1_EXO")

#extract TPM of EXO marker genes early
final.dat.exo1 = matrix(ncol = 7, nrow = 0)

for (i in 1:length(marker.genes.exo)) {
  m.gene.pos = which(dat.tc.ref[marker.genes.exo[i], ] > 0)
  if(length(m.gene.pos) > 0) {
    dat.gene = tsn.tc.Vc.iN1[which(tsn.tc.Vc.iN1$Cell %in% colnames(dat.tc.ref[, m.gene.pos, drop = F])), ] %>%
      mutate(Gene = gsub("_EXO", "", marker.genes.exo[i]),
             TPM = as.numeric(dat.tc.ref[marker.genes.exo[i], which(colnames(dat.tc.ref) %in% .$Cell)]))
    final.dat.exo1 = rbind(final.dat.exo1, dat.gene)}}

#scale
final.dat.exo1 = final.dat.exo1 %>%
  as.data.table %>%
  .[order(Gene, TPM)] %>%
  mutate(TPM = .[, scale(TPM), by = Gene]$V1)

#plot EXO data in facet grid early
p = ggplot(tsn.tc.Vc.iN1, aes(tSNE_1, tSNE_2)) +
  geom_point(size = 3, shape = 21, fill = "grey30", color = "grey10") +
  geom_point(data = final.dat.exo1, aes(tSNE_1, tSNE_2, fill = TPM, color = TPM), size = 3, shape = 21) +
  scale_fill_gradient2(limits = c(min(final.dat.exo1$TPM), max(final.dat.exo1$TPM)), 
                       low = "grey30", 
                       mid = "white", 
                       high = "#e99400", 
                       midpoint = (0)) +
  scale_color_gradient2(limits = c(min(final.dat.exo1$TPM), max(final.dat.exo1$TPM)), 
                        low = "grey30", 
                        mid = "white", 
                        high = "#e99400", 
                       midpoint = (-0.5)) +
  scale_x_continuous(breaks = c(-7, 5, 17)) +
  theme_jo() +
  theme(axis.text = element_blank(),
        strip.text = element_text(size = 22),
        panel.border = element_rect(fill = NA, size = 1.5, color = "grey10"),
        axis.title = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank()) +
  facet_wrap(~Gene, ncol = 4)
subchunkify(p, 12, 8)

#extract TPM of EXO marker genes late
final.dat.exo2 = matrix(ncol = 7, nrow = 0)

for (i in 1:length(marker.genes.exo)) {
  m.gene.pos = which(dat.tc.ref[marker.genes.exo[i], ] > 0)
  
  if(length(m.gene.pos) > 0) {
    dat.gene = tsn.tc.Vc.iN2[which(tsn.tc.Vc.iN2$Cell %in% colnames(dat.tc.ref[ , m.gene.pos, drop=F])), ] %>%
      mutate(Gene = gsub("_EXO", "", marker.genes.exo[i]), 
             TPM = as.numeric(dat.tc.ref[marker.genes.exo[i], which(colnames(dat.tc.ref) %in% .$Cell)]))
    final.dat.exo2 = rbind(final.dat.exo2, dat.gene)}}

#scale
final.dat.exo2 = final.dat.exo2 %>%
  as.data.table %>%
  .[order(Gene, TPM)] %>%
  mutate(TPM = .[, scale(TPM), by = Gene]$V1)

#plot EXO data in facet grid late
p = ggplot(tsn.tc.Vc.iN2, aes(tSNE_1, tSNE_2)) +
  geom_point(size = 3, shape = 21, fill = "grey30", color = "grey10") +
  geom_point(data = final.dat.exo2, aes(tSNE_1, tSNE_2, fill = TPM, color = TPM), size = 3, shape = 21) +
  scale_fill_gradient2(limits = c(min(final.dat.exo2$TPM), max(final.dat.exo2$TPM)), 
                       low = "grey30", 
                       mid = "white", 
                       high = "#e99400", 
                       midpoint = (0)) +
  scale_color_gradient2(limits = c(min(final.dat.exo2$TPM), max(final.dat.exo2$TPM)), 
                        low = "grey30", 
                        mid = "white", 
                        high = "#e99400", 
                       midpoint = (-0.5)) +
  scale_y_continuous(breaks = c(-10, 0, 10)) +
  theme_jo() +
  theme(axis.text = element_blank(),
        strip.text = element_text(size = 22),
        panel.border = element_rect(fill = NA, size = 1.5, color = "grey10"),
        axis.title = element_blank(),
        axis.line = element_blank(),
        axis.ticks = element_blank()) +
  facet_wrap(~Gene, ncol = 4)
subchunkify(p, 12, 8)
```

##########################################################
#################        FIGURE S5       #################
##########################################################

```{r FIGURES5, warning = FALSE, message = FALSE, results = 'asis'}
#----------------------------FigureS5A--------------------------------
###################### Seurat ######################
#read Time-course data, create Seurat
dat.ref.tc.count = readat("time.course.count.mat.qc.csv")
tc.data = Matrix(as.matrix(dat.ref.tc.count), sparse = T)

#seurat
tcs.all = CreateSeuratObject(raw.data = tc.data, min.cells = 3, normalization.method = "LogNormalize", display.progress = F) %>%
  FindVariableGenes(x.low.cutoff = 0.01, x.high.cutoff = 8, y.cutoff = 1, do.plot = F, display.progress = F) %>%
  ScaleData(display.progress = F) %>%
  RunPCA(do.print = F, pcs.print = 0, genes.print = 0) %>%
  ProjectPCA(do.print = FALSE, pcs.print = 0, genes.print = 0) %>%
  FindClusters(dims.use = 1:7, resolution = 1, print.output = F) %>%
  RunTSNE(dims.use = 1:7, do.fast = TRUE)
#TSNEPlot(object = tcs.all, do.label = T)  

#create data frame for ggplot
tsn.all = as.data.frame(tcs.all@dr$tsne@cell.embeddings) %>%
  rownames_to_column(var = "Cell") %>%
  mutate(Sample = unlist(lapply(strsplit(Cell, "[.]"), function(X) X[[1]])),
         Cluster = factor(tcs.all@meta.data$res.1, levels = c(2, 3, 4, 1, 0)))

###################### Monocle ######################
iNM_sample_sheet = readat("time.course.sample.sheet.csv")
iNM_gene_annotation = readat("time.course.gene.annotation.csv")
iNM_expr_matrix = readat("time.course.tpm.mat.qc.csv")
row.names(iNM_gene_annotation) = row.names(iNM_expr_matrix)

#create cell data set
pd = new("AnnotatedDataFrame", data = iNM_sample_sheet)
fd = new("AnnotatedDataFrame", data = iNM_gene_annotation)
iNM = newCellDataSet(as.matrix(iNM_expr_matrix), phenoData = pd, featureData = fd)

#filter Genes
iNM = detectGenes(iNM,min_expr = 0.1)
expressed_genes = row.names(subset(fData(iNM), num_cells_expressed >= 3))

#diff.test = differentialGeneTest(iNM[expressed_genes, ], fullModelFormulaStr = "~Media", cores = 4)
#sig_genes = subset(diff.test, qval < 0.01)
#ordering_genes = sig_genes$gene_short_name[which(sig_genes$gene_short_name %in% expressed_genes)]

#obtain nervous and blood ordering genes
ordering_genes = readat("time.course.diff.genes.media.csv") %>%
  .$Gene

ordering_genes = ordering_genes[which(ordering_genes %in% expressed_genes)]

#reduce dimension, order cells
iNM = setOrderingFilter(iNM, ordering_genes) %>%
  reduceDimension(reduction_method = "DDRTree") %>%
  orderCells(reverse = T)

#obtain data frame
pt.iN = plot_cell_trajectory(iNM, color_by = "Pseudotime", show_tree = T, show_branch_points = T)
pt.dat = pt.iN$data %>%
  column_to_rownames(var = "sample_name") %>%
  .[row.names(as.data.frame(t(iNM@reducedDimS))), ]

#extract information for transferring data to ggplot2
iNM.coord = as.data.frame(t(iNM@reducedDimS)) %>%
  mutate(Batch = unlist(lapply(strsplit(row.names(.), "[.]"), function(X) X[[1]])),
         Cell = row.names(as.data.frame(t(iNM@reducedDimS))),
         State = pt.dat$State,
         Sample = factor(gsub2(c("TFi2", "TFi4"), c("TFi1", "TFi3"), Batch), levels = c("FIB", "Ci1", "TFi1", "Ci2", "TFi3"))) %>%
  left_join(., pt.iN$data[ ,c("sample_name", "Pseudotime")], by = c("Cell" = "sample_name"))

#obtain minimum spanning tree
#piN
reduced_dim_coords = reducedDimK(iNM)
ica_space_df = data.frame(Matrix::t(reduced_dim_coords[c(1, 2), ])) %>%
  select(prin_graph_dim_1 = X1, prin_graph_dim_2 = X2) %>%
  rownames_to_column(var = "sample_name") %>%
  mutate(sample_state = sample_name)
iN.mst = minSpanningTree(iNM)
edge_list = as.data.frame(igraph::get.edgelist(iN.mst)) %>%
  select(source = V1, target = V2)
edge_df_iN = full_join(ica_space_df, edge_list, by = c("sample_name" = "source")) %>%
  rename(c("prin_graph_dim_1" = "source_prin_graph_dim_1", "prin_graph_dim_2" = "source_prin_graph_dim_2")) %>%
  full_join(ica_space_df[ ,c("sample_name", "prin_graph_dim_1", "prin_graph_dim_2")], by = c("target" = "sample_name")) %>%
  rename(c("prin_graph_dim_1" = "target_prin_graph_dim_1", "prin_graph_dim_2" = "target_prin_graph_dim_2"))

#create color palettes
my.pal.tc.monoc1 = c("FIB" = "#033782", "Ci1" = "#8b008b", "Ci2" = "#c71585", 
                     "TFi1" = "#ee2c2c", "TFi2" = "#ee2c2c", "TFi3" = "#ee7600", "TFi4" = "#ee7600")
my.pal.tc.monoc2 = c("FIB" = "#011736","Ci1" = "#400040", "Ci2" = "#7a0d52", 
                     "TFi1" = "#a11d1d", "TFi2" = "#a11d1d", "TFi3" = "#a15000", "TFi4" = "#a15000")

#plot with sample information
p = ggplot() +
  geom_segment(aes_string(x = "source_prin_graph_dim_1", y = "source_prin_graph_dim_2", xend = "target_prin_graph_dim_1",
                          yend = "target_prin_graph_dim_2"), size = 1.2, linetype = "solid", na.rm = TRUE, data = edge_df_iN) +
  geom_point(data = iNM.coord[sample(nrow(iNM.coord)), ], aes(V1, V2, fill = Batch, color = Batch), size = 7, shape = 21) +
  scale_fill_manual(values = my.pal.tc.monoc1) +
  scale_color_manual(values  =my.pal.tc.monoc2) +
  theme_jo() +
  theme(panel.border = element_rect(fill = NA, size = 1.5, color = "grey10"),
        axis.line = element_blank())
subchunkify(p, 6, 6)

#----------------------------FigureS5B--------------------------------
#use protein coding genes that are expressed in at least 20 cells
expressed_genes = row.names(subset(fData(iNM), num_cells_expressed >= 20))
genes = readat("gene.class.csv") %>%
  filter(Class == "protein_coding") %>%
  .$Gene %>%
  .[which(. %in% expressed_genes)]

#BEAM
BEAM.res = BEAM(iNM[genes, ], branch_point = 4, cores = 4)
BEAM.res.ord = BEAM.res[order(BEAM.res$qval), ] %>%
  select(c("gene_short_name", "pval", "qval"))

#produce heatmap
beam.tc = plot_genes_branched_heatmap(iNM[row.names(subset(BEAM.res.ord, qval < 1e-8)),], 
                                      branch_point = 4, 
                                      norm_method = "log", 
                                      return_heatmap = T,
                                      num_clusters = 6)

beam.mat = as.data.frame(beam.tc[[3]])
ph = pheatmap(beam.mat, cluster_cols = F, cutree_rows = 6, clustering_method = "ward.D", fontsize_row = 2)

#create annotation
dat.dend = as.dendrogram(ph$tree_row)
my.clust = as.hclust(dat.dend)
annot.col = beam.tc$annotation_col %>%
  rename(c("Cell Type" = "Celltype"))

#colors
ann.colors = list(Celltype = c("Cell fate 1" = "#9d9d9f", "Pre-branch" = "#4575b4", "Cell fate 2" = "#333333"))

#plot heatmap
p = pheatmap(beam.mat, cluster_rows = my.clust, cluster_cols = F, annotation_col = annot.col,
         annotation_colors = ann.colors, show_rownames = F, show_colnames = F, annotation_legend = F, fontsize_row = 1, 
         border_color = NA, cutree_rows = 6, fontsize = 20, clustering_method = "ward.D2", legend = F)
subchunkify(p, 10, 6)

#----------------------------FigureS5D--------------------------------
#Pseudotime-dependent genes
#differential gene expression along pseudotime
diff_test_res.pt = differentialGeneTest(iNM[genes,], fullModelFormulaStr = "~sm.ns(Pseudotime)", cores = 4)
sig_gene_names = row.names(subset(diff_test_res.pt, qval < 1e-4))

#scale matrix
cds_subset = iNM[sig_gene_names,]
newdata = data.frame(Pseudotime = seq(min(pData(cds_subset)$Pseudotime), max(pData(cds_subset)$Pseudotime),length.out = 100))
m = genSmoothCurves(cds_subset, cores = 1, trend_formula = "~sm.ns(Pseudotime, df = 3)",  relative_expr = T, new_data = newdata)
m = m[!apply(m, 1, sum) == 0,]
m = log10(m + 1)
m = m[!apply(m, 1, sd) == 0, ]
m = Matrix::t(scale(Matrix::t(m), center = TRUE))
m = m[is.na(row.names(m)) == FALSE, ]
m[is.nan(m)] = 0
m[m > 3] = 3
m[m < -3] = -3
pt.matrix = m

#create pheatmap object
ph = pheatmap(pt.matrix, cluster_cols = F, cutree_rows = 3, clustering_method = "ward.D2", fontsize_row = 1)

#create color annotation
annot.row = as.data.frame(cutree(ph$tree_row, k = 3)) %>%
  rownames_to_column(var = "Gene")
colnames(annot.row) = c("Gene", "Cluster")
annot.row$Cluster = chartr("123", "ABC", annot.row$Cluster)
dat.dend = as.dendrogram(ph$tree_row) %>%
  rotate(annot.row[order(annot.row$Cluster), ]$Gene)
annot.row = annot.row %>%
  column_to_rownames(var = "Gene")
my.clust = as.hclust(dat.dend)

ann.colors = list(Cluster = c("A" = "#5aaab1", "B" = "#fda24a", "C" = "#1a5b87"))

#plot heatmap
p = pheatmap(pt.matrix, cluster_rows = my.clust, cluster_cols = F, annotation_row = annot.row, annotation_colors = ann.colors, 
         show_rownames = F, show_colnames = F, annotation_legend = F, border_color = NA, fontsize_row = 0.5, cutree_rows = 3,
         fontsize = 20, clustering_method = "ward.D2", legend = F)
subchunkify(p, 10, 6)

#----------------------------FigureS5F-H--------------------------------
#plot genes in pseudotime
#create data frame
dat.pt = iNM.coord %>%
  .[order(.$Pseudotime), ]
  
#set marker genes
m.genes.fib = c("CCNB1", "MKI67", "TK1", "TOP2A")
m.genes.neuro = c("NRCAM", "SFRP1", "SNAP25", "SYT1")
m.genes.blood = c("COL15A1", "THSD7A", "HAS2", "PTGS2")

#create data frames for plotting
#-------------------------------Fibroblast genes------------------------
#create in data frame for plotting
dat.pt.final.fib = matrix(nrow = 0, ncol = 4)

for (i in seq_len(length(m.genes.fib))) {
  
  dat.marker.fib = as.data.frame(t(dat.tc.ref[m.genes.fib[i], ])) %>%
    mutate(Sample = iNM.coord$Sample,
           Pseudotime = iNM.coord$Pseudotime,
           Gene = m.genes.fib[i])
  names(dat.marker.fib) = c("Tpm", "Sample", "Pseudotime", "Gene")
  dat.pt.final.fib = rbind(dat.pt.final.fib, dat.marker.fib)
}

#-------------------------------Neuronal genes--------------------------
#create in data frame for plotting
dat.pt.final.neuro = matrix(nrow = 0, ncol = 4)

for (i in seq_len(length(m.genes.neuro))) {
  
  dat.marker.neuro = as.data.frame(t(dat.tc.ref[m.genes.neuro[i], ])) %>%
    mutate(Sample = iNM.coord$Sample,
           Pseudotime = iNM.coord$Pseudotime,
           Gene = m.genes.neuro[i])
  names(dat.marker.neuro) = c("Tpm", "Sample", "Pseudotime", "Gene")
  dat.pt.final.neuro = rbind(dat.pt.final.neuro, dat.marker.neuro)
}

#-------------------------------Vascular genes--------------------------
#create in data frame for plotting
dat.pt.final.blood = matrix(nrow = 0, ncol = 4)

for (i in seq_len(length(m.genes.blood))) {
  
  dat.marker.blood = as.data.frame(t(dat.tc.ref[m.genes.blood[i], ])) %>%
    mutate(Sample = iNM.coord$Sample,
           Pseudotime = iNM.coord$Pseudotime,
           Gene = m.genes.blood[i])
  names(dat.marker.blood) = c("Tpm", "Sample", "Pseudotime", "Gene")
  dat.pt.final.blood = rbind(dat.pt.final.blood, dat.marker.blood)
}

#modify marker levels
dat.pt.final.fib$Gene = factor(dat.pt.final.fib$Gene, levels = c("CCNB1", "MKI67", "TK1", "TOP2A"))
dat.pt.final.fib$Gene = factor(dat.pt.final.fib$Gene, levels = c("CCNB1", "MKI67", "TK1", "TOP2A"))

#create color palette
mypal.pt1=c("FIB" = "#033782", "Ci1" = "#8b008b", "Ci2" = "#c71585", "TFi1" = "#ee2c2c", "TFi3" = "#ee7600")
mypal.pt2=c("FIB" = "#011736","Ci1" = "#400040", "Ci2" = "#7a0d52", "TFi1" = "#a11d1d", "TFi3" = "#a15000")

#----------------------------FigureS5F---------------------------------
#plot
p = ggplot(dat.pt.final.fib, aes(Pseudotime, Tpm)) +
  geom_point(aes(color = as.factor(Sample), fill = as.factor(Sample)), size = 5, alpha = 0.8, shape = 21) +
  scale_color_manual(values = mypal.pt2) +
  scale_fill_manual(values = mypal.pt1) +
  geom_smooth(se = T, size = 2, color = "grey30", span = 0.5) +
  ylab("Log2(TPM)") +
  theme_jo()+
  theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank()) +
  facet_wrap(~Gene, ncol = 1, scales = "free")
subchunkify(p, 10, 8)

#----------------------------FigureS5G--------------------------------
#plot
p = ggplot(dat.pt.final.neuro, aes(Pseudotime, Tpm)) +
  geom_point(aes(color = as.factor(Sample), fill = as.factor(Sample)), size = 5, alpha = 0.8, shape = 21) +
  scale_color_manual(values = mypal.pt2) +
  scale_fill_manual(values = mypal.pt1) +
  geom_smooth(se = T, size = 2, color = "grey30", span = 0.5) +
  ylab("Log2(TPM)") +
  theme_jo()+
  theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank()) +
  facet_wrap(~Gene, ncol = 1, scales = "free")
subchunkify(p, 10, 8)

#----------------------------FigureS5H--------------------------------
#plot
p = ggplot(dat.pt.final.blood, aes(Pseudotime, Tpm)) +
  geom_point(aes(color = as.factor(Sample), fill = as.factor(Sample)), size = 5, alpha = 0.8, shape = 21) +
  scale_color_manual(values = mypal.pt2) +
  scale_fill_manual(values = mypal.pt1) +
  geom_smooth(se = T, size = 2, color = "grey30", span = 0.5) +
  ylab("Log2(TPM)") +
  theme_jo()+
  theme(axis.ticks.x = element_blank(),
        axis.text.x = element_blank()) +
  facet_wrap(~Gene, ncol = 1, scales = "free")
subchunkify(p, 10, 8)
```

##########################################################
#################        FIGURE 5        #################
##########################################################

```{r FIGURE5, warning = FALSE, message = FALSE, results = 'asis'}
#----------------------------Figure5A--------------------------------
###################### Seurat ######################
#read Time-course data, create Seurat
dat.ref.tc.count = readat("time.course.count.mat.qc.csv")
tc.data = Matrix(as.matrix(dat.ref.tc.count), sparse = T)

#seurat
tcs.all = CreateSeuratObject(raw.data = tc.data, min.cells = 3, normalization.method = "LogNormalize", display.progress = F) %>%
  FindVariableGenes(x.low.cutoff = 0.01, x.high.cutoff = 8, y.cutoff = 1, do.plot = F, display.progress = F) %>%
  ScaleData(display.progress = F) %>%
  RunPCA(do.print = F, pcs.print = 0, genes.print = 0) %>%
  ProjectPCA(do.print = FALSE, pcs.print = 0, genes.print = 0) %>%
  FindClusters(dims.use = 1:7, resolution = 1, print.output = F) %>%
  RunTSNE(dims.use = 1:7, do.fast = TRUE)
#TSNEPlot(object = tcs.all, do.label = T)  

#create data frame for ggplot
tsn.all = as.data.frame(tcs.all@dr$tsne@cell.embeddings) %>%
  rownames_to_column(var = "Cell") %>%
  mutate(Sample = unlist(lapply(strsplit(Cell, "[.]"), function(X) X[[1]])),
         Cluster = factor(tcs.all@meta.data$res.1, levels = c(2, 3, 4, 1, 0)))

###################### Monocle ######################
iNM_sample_sheet = readat("time.course.sample.sheet.csv")
iNM_gene_annotation = readat("time.course.gene.annotation.csv")
iNM_expr_matrix = readat("time.course.tpm.mat.qc.csv")
row.names(iNM_gene_annotation) = row.names(iNM_expr_matrix)

#create cell data set
pd = new("AnnotatedDataFrame", data = iNM_sample_sheet)
fd = new("AnnotatedDataFrame", data = iNM_gene_annotation)
iNM = newCellDataSet(as.matrix(iNM_expr_matrix), phenoData = pd, featureData = fd)

#filter Genes
iNM = detectGenes(iNM,min_expr = 0.1)
expressed_genes = row.names(subset(fData(iNM), num_cells_expressed >= 3))

#obtain nervous and blood ordering genes
ordering_genes = readat("genes.nervous.blood.csv") %>%
  .$Gene

#reduce dimension, order cells
iNM = setOrderingFilter(iNM, ordering_genes) %>%
  reduceDimension(reduction_method = "DDRTree") %>%
  orderCells(reverse = F)

#obtain data frame
pt.iN = plot_cell_trajectory(iNM, color_by = "State", show_tree = T, show_branch_points = T)
pt.dat = pt.iN$data %>%
  column_to_rownames(var = "sample_name") %>%
  .[row.names(as.data.frame(t(iNM@reducedDimS))), ]

#extract information for transferring data to ggplot2
iNM.coord = as.data.frame(t(iNM@reducedDimS)) %>%
  mutate(Sample = unlist(lapply(strsplit(row.names(.), "[.]"), function(X) X[[1]])),
         Cell = row.names(as.data.frame(t(iNM@reducedDimS))),
         State = pt.dat$State,
         Branch = chartr("1234567", "AABBBCA", State)) %>%
  left_join(., pt.iN$data[ ,c("sample_name", "Pseudotime")], by = c("Cell" = "sample_name"))

#obtain minimum spanning tree
#piN
reduced_dim_coords = reducedDimK(iNM)
ica_space_df = data.frame(Matrix::t(reduced_dim_coords[c(1, 2), ])) %>%
  select(prin_graph_dim_1 = X1, prin_graph_dim_2 = X2) %>%
  rownames_to_column(var = "sample_name") %>%
  mutate(sample_state = sample_name)
iN.mst = minSpanningTree(iNM)
edge_list = as.data.frame(igraph::get.edgelist(iN.mst)) %>%
  select(source = V1, target = V2)
edge_df_iN = full_join(ica_space_df, edge_list, by = c("sample_name" = "source")) %>%
  rename(c("prin_graph_dim_1" = "source_prin_graph_dim_1", "prin_graph_dim_2" = "source_prin_graph_dim_2")) %>%
  full_join(ica_space_df[ ,c("sample_name", "prin_graph_dim_1", "prin_graph_dim_2")], by = c("target" = "sample_name")) %>%
  rename(c("prin_graph_dim_1" = "target_prin_graph_dim_1", "prin_graph_dim_2" = "target_prin_graph_dim_2"))

#create color palettes
my.pal.tc.monoc1=c("FIB" = "#033782", "Ci1" = "#8b008b", "Ci2" = "#c71585", "TFi1" = "#ee2c2c", "TFi2" = "#ee2c2c", 
                   "TFi3" = "#ee7600", "TFi4" = "#ee7600")
my.pal.tc.monoc2=c("FIB" = "#011736","Ci1" = "#400040", "Ci2" = "#7a0d52", "TFi1" = "#a11d1d", "TFi2" = "#a11d1d", 
                   "TFi3" = "#a15000", "TFi4" = "#a15000")

#plot with sample information
p = ggplot() +
  geom_segment(aes_string(x = "source_prin_graph_dim_1", y = "source_prin_graph_dim_2", xend = "target_prin_graph_dim_1",
                          yend = "target_prin_graph_dim_2"), size = 1.2, linetype = "solid", na.rm = TRUE, data = edge_df_iN) +
  geom_point(data = iNM.coord[sample(nrow(iNM.coord)), ], aes(V1, V2, fill = Sample, color = Sample), size = 7, shape = 21) +
  scale_fill_manual(values = my.pal.tc.monoc1) +
  scale_color_manual(values  =my.pal.tc.monoc2) +
  theme_jo() +
  theme(panel.border = element_rect(fill = NA, size = 1.5, color = "grey10"))
subchunkify(p, 6, 6)

#----------------------------Figure5B--------------------------------
#read.data
dat.tc.ref = readat("time.course.tpm.mat.qc.log.csv")

#define marker genes
marker.genes = c("CHRNA1", "GABRB2", "TAC3", "KCNJ12", "MKI67", "TK1", 
                 "CDC20", "KIF20A", "FSTL3", "FSTL5", "ISM1", "PTH1R")

# Extract TPM of EXO marker genes
final.dat = matrix(ncol = 7, nrow = 0)

for (i in 1:length(marker.genes)) {
  m.gene.pos= which(dat.tc.ref[marker.genes[i], ] > 0)
  if(length(m.gene.pos) > 0) {
    dat.gene = iNM.coord[which(iNM.coord$Cell %in% colnames(dat.tc.ref[ ,m.gene.pos, drop = F])), ] %>%
      mutate(Gene = marker.genes[i],
             TPM = as.numeric(dat.tc.ref[marker.genes[i], which(colnames(dat.tc.ref) %in% .$Cell)]))
    final.dat = rbind(final.dat, dat.gene)}}

#scale
final.dat = final.dat %>%
  mutate(Gene = factor(Gene, levels = marker.genes)) %>%
  as.data.table %>%
  .[order(Gene, TPM)] %>%
  mutate(TPM = .[, scale(TPM), by = Gene]$V1)

#Plot EXO data in facet grid
p = ggplot(iNM.coord, aes(V1, V2)) +
  geom_point(size = 3.5, shape = 21, fill = "grey30", color = "grey10") +
  geom_point(data = final.dat, aes(V1, V2, fill = TPM, color = TPM), size = 3.5, shape = 21) +
  scale_fill_gradient2(limits = c(min(final.dat$TPM), max(final.dat$TPM)), 
                       low="grey30", 
                       mid="white", 
                       high="#e99400", 
                       midpoint = (max(final.dat$TPM)/4)) +
  scale_color_gradient2(limits = c(min(final.dat$TPM), max(final.dat$TPM)), 
                       low="grey30", 
                       mid="white", 
                       high="#e99400", 
                       midpoint = (max(final.dat$TPM)/6)) +
  theme_jo() +
  theme(axis.ticks = element_blank(),
        axis.text = element_blank(),
        axis.title = element_blank()) +
  facet_wrap(~Gene, scales = "free")
subchunkify(p, 8, 10)

#----------------------------Figure5D-------------------------------
#define exogenous TFs
marker.genes.exo = c("ASCL1_EXO", "DLX1_EXO", "DLX2_EXO", "FEV_EXO", "FOXA2_EXO", "FOXP2_EXO", "ISL1_EXO", "LHX2_EXO", 
                     "NEUROD1_EXO", "NEUROG2_EXO", "NR2F1_EXO", "NR2F2_EXO", "NR4A2_EXO", "OLIG2_EXO", "OTX2_EXO", 
                     "PAX6_EXO", "PITX3_EXO", "POU3F2_EXO", "TLX3_EXO", "ZIC1_EXO")

#create data frame
pt.dat = pt.iN$data %>%
  mutate(State = as.numeric(as.character(State)),
         Branch = chartr("1234567", "AABBBCA", State))
pt.branch = pt.dat %>%
  filter(Branch %in% c("B", "C")) %>%
  select(c("sample_name","Branch"))
pt.split = pt.branch %>%
  split(.$Branch)
dat.split = lapply(pt.split, function(X) dat.tc.ref[grep("_EXO", row.names(dat.tc.ref)), X$sample_name])
  
#perform fisher tests
my.fisher.pos = matrix(nrow = 2, ncol = 0)

for (i in 1:length(marker.genes.exo)){
  exo = marker.genes.exo[i]
  exo.pos = length(which(dat.tc.ref[exo, pt.branch$sample_name] > 0))
  num.cells = length(row.names(pt.branch))
  # Perform Fisher exact test
  x.list = lapply(dat.split, function(x) length(which(x[exo, ] > 0)))
  m.list = mapply(function(x,y) length(colnames(x))-y, dat.split, x.list, SIMPLIFY = F)
  n.list = lapply(x.list, function(x) exo.pos - x)
  k.list = mapply(function(x, y) num.cells - length(colnames(x)) - y, dat.split, n.list, SIMPLIFY = F)
  my.fisher = mapply(function(a, b, c, d) fisher.test(matrix(c(a, b, c, d), nrow = 2), alternative = "greater"),
                     x.list, m.list, n.list, k.list, SIMPLIFY = F)
  my.p = do.call("rbind", (lapply(my.fisher, function(x) x$p.value)))
  colnames(my.p) = exo
  my.fisher.pos = cbind(my.fisher.pos, my.p)
}

#modify fisher results
my.fisher.final = as.data.frame(my.fisher.pos) %>%
  mutate(Branch = c("B", "C")) %>%
  melt(id = "Branch")
my.fisher.final$value = -log10(my.fisher.final$value)

#create final data frame
final.branch = iNM.coord %>%
  filter(Branch %in% c("B", "C")) %>%
  right_join(my.fisher.final, by = "Branch") %>%
  mutate(EXO = gsub("_EXO", "", variable))

#create color annotation
my.vec = vector(mode = "character", length = length(row.names(final.branch)))
my.vec[which(final.branch$value >= 0 & final.branch$value <= 1.30103)] = "A"
my.vec[which(final.branch$value > 1.30103 & final.branch$value <= 3)] = "B"
my.vec[which(final.branch$value > 3 & final.branch$value <= 5)] = "C"
my.vec[which(final.branch$value > 5)] = "D"
final.branch$Color = my.vec

#create color palette
my.pal.sig = c("A" = "white", "B" = "#f9d300", "C" = "#fb7c00", "D" = "#ed2323")

#plot early
p = ggplot()+
  geom_point(data = iNM.coord, aes(V1, V2), size = 3.5, shape = 21, fill = "white", color = "grey20") +
  geom_point(data = final.branch, aes(V1, V2, fill=Color), color = "grey20", shape = 21, size = 3.5) +
  scale_fill_manual(values = my.pal.sig) +
  theme_blank() +
  theme(axis.ticks = element_blank(),
        axis.line = element_line(size = 1)) +
  facet_wrap(~EXO, nrow = 4, scales = "free")
subchunkify(p, 8, 10)
```